docs/adr/0016-use-graphql-and-apollo-for-settings-redesign.md

```diff
@@ -0,0 +1,138 @@
+
+- Queries always return a 200 status code, even if the query errored. If the query is unsuccessful, the response will instead have an `errors` key with associated error messages and stacktrace. However, the error messaging can be quite detailed and include all the resolvers and refer to the exact part of the query at fault.
+
+##### Cons
+
+- The learning curve that comes with introducing new tech.
+- Initial setup time of another service (`graphql-api-server`).
+- While GQL caching can be done at the database or client level though Apollo, GQL doesn't rely on HTTP caching methods.
+- While GQL has been around long enough to gain excellent community support, REST has been around with established patterns for even longer.
+
+### Option B - Layer GraphQL and Apollo on top of our REST architecture, making direct database calls for basic operations
+
+#### Description
+
+- This option allows us to begin working with GraphQL and Apollo on top of our REST architecture, proxying existing auth-server calls, except where it makes sense to interact with the database directly for simple operations.
+
+##### Pros:
+
+- Less risk and less upfront work than option C.
+- We'll still heavily reduce the number of initial network requests performed on the Settings frontend with an incremental approach. The Apollo Server can orchestrate the REST requests sufficient to fulfil just that query and return exactly the data requested.
+- Apollo server can proxy to existing auth-server calls. We can incrementally swap out auth-server API calls with direct implementation.
+
+##### Cons:
+
+- Creates tech debt. We won't be fully integrated until we finish swapping out the auth-server API calls.
+- Our stack will be less universal across FxA. We'll use REST for non-Settings FxA pages and for the payments server until refactoring can be done.
+- While GraphQL is database agnostic, we will be unable to use [GraphQL Subscriptions](https://www.apollographql.com/docs/react/data/subscriptions/) because our database does not support real-time data.
+
+### Option C - Full GraphQL and Apollo integration with direct database calls
+
+#### Description
+
+- This option would extend option B. The more complex operations routed through the auth-server would be directly implemented through the Apollo server.
+
+##### Pros
+
+- Less tech debt later.
+- We would see a performance boost over option B.
+- An all-out dedication to this approach could open the door to evaluating our database options (and being able to work with [GraphQL subscriptions](https://www.apollographql.com/docs/react/data/subscriptions/)) or more quickly pave a path to move our backend off of MySQL to something else for a more regional conscious architecture.
+
+##### Cons
+
+- Increased risk. While we'll be pushing the new Settings app to staging and production while we're building it out, the Settings Redesign project already has a lot of moving parts and an incremental approach definitely feels more ideal.
```

Comment:
I'd say it's worth further noting the risk here of having two different systems hitting the same database: Existing REST endpoints in auth-server aren't just queries, they embed business logic. We'd have to duplicate that logic in GraphQL and test in tandem with REST to make sure we don't break anything on either side.

My concern here is that the tangle between existing settings-with-REST and new settings-with-GraphQL could stomp over a lot of the productivity gains of adopting GraphQL in the first place. 

I might be over-worried about that, and maybe it's a quick rewrite. But surprises and the second-system effect is why folks so often [advise against greenfield rewrites](https://www.joelonsoftware.com/2000/04/06/things-you-should-never-do-part-i/).